Change type of timeoutFutures_ key to time_point instead of duration

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30355 Tweak pollTimedOutRPCs thread synchronization
* **#31078 Change type of timeoutFutures_ key to time_point instead of duration**

Make `ProcessGroupAgent::pollTimedOutRPCs` code more conventional.

- Use `std::chrono::time_point` to represent `endTime` instead of `std::chrono::duration`.
- Replace `std::condition_variable::wait_for(lock, endTime)` with `std::condition_variable::wait_until(lock, endTime)`.
- Remove the unnecessary `::getRPCRemainingTime()`.

Differential Revision: [D5705442](https://our.internmc.facebook.com/intern/diff/D5705442/)

xush6528 committed:
Change type of timeoutFutures_ key to time_point instead of duration

Make `ProcessGroupAgent::pollTimedOutRPCs` code more conventional.

- Use `std::chrono::time_point` to represent `endTime` instead of `std::chrono::duration`.
- Replace `std::condition_variable::wait_for(lock, entTime)` with `std::condition_variable::wait_until(lock, entTime)`.
- Remove the unnecessary `::getRPCRemainingTime()`.

Differential Revision: [D5705442](https://our.internmc.facebook.com/intern/diff/D5705442/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -472,19 +466,16 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
       futureTimeoutCV_.wait(lock);
     } else {
-      futureTimeoutCV_.wait_for(lock, sleepTime);
+      futureTimeoutCV_.wait_until(lock, minEndTime);
```
What happens if `minEndTime < now()`? I think we'd want the thread to wake up immediately and begin processing timed out futures, is this the case?

xush6528 replied:
In that case, this call does not block.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -77,6 +77,12 @@ class ProcessGroupAgent : public RpcAgent {
+  using steady_clock_time_point =
+      std::chrono::time_point<std::chrono::steady_clock>;
+
+  static constexpr steady_clock_time_point kInfiniteTimeoutTimePoint =
```
Should this go in rpc_agent.h so that other RpcAgents can have access to them?

xush6528 replied:
I don't think so. This watch dog thread and using a const time to represent infinity are all specific to ProcessGroupAgent only.

Notice this is not the "Default RPC timeout".

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -472,19 +466,16 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
       futureTimeoutCV_.wait(lock);
     } else {
-      futureTimeoutCV_.wait_for(lock, sleepTime);
+      futureTimeoutCV_.wait_until(lock, minEndTime);
```
I think you can get rid of the `sleepTime` variable now?

xush6528 replied:
Right. Let me update.

mrshenli reviewed (approved):
LGTM! Please get a stamp from @rohan-varma as well.

There is typo in the PR description `entTime` -> `endTime`. Please make sure the final commit message has the correct spelling. Thanks!

xush6528 committed:
Update on "Change type of timeoutFutures_ key to time_point instead of duration"


Make `ProcessGroupAgent::pollTimedOutRPCs` code more conventional.

- Use `std::chrono::time_point` to represent `endTime` instead of `std::chrono::duration`.
- Replace `std::condition_variable::wait_for(lock, entTime)` with `std::condition_variable::wait_until(lock, entTime)`.
- Remove the unnecessary `::getRPCRemainingTime()`.

Differential Revision: [D5705442](https://our.internmc.facebook.com/intern/diff/D5705442/)

[ghstack-poisoned]

xush6528 commented:
@mrshenli  Thanks! updated PR description.

xush6528 committed:
Update on "Change type of timeoutFutures_ key to time_point instead of duration"


Make `ProcessGroupAgent::pollTimedOutRPCs` code more conventional.

- Use `std::chrono::time_point` to represent `endTime` instead of `std::chrono::duration`.
- Replace `std::condition_variable::wait_for(lock, endTime)` with `std::condition_variable::wait_until(lock, endTime)`.
- Remove the unnecessary `::getRPCRemainingTime()`.

Differential Revision: [D5705442](https://our.internmc.facebook.com/intern/diff/D5705442/)

[ghstack-poisoned]

xush6528 committed:
Update on "Change type of timeoutFutures_ key to time_point instead of duration"


Make `ProcessGroupAgent::pollTimedOutRPCs` code more conventional.

- Use `std::chrono::time_point` to represent `endTime` instead of `std::chrono::duration`.
- Replace `std::condition_variable::wait_for(lock, endTime)` with `std::condition_variable::wait_until(lock, endTime)`.
- Remove the unnecessary `::getRPCRemainingTime()`.

Differential Revision: [D5705442](https://our.internmc.facebook.com/intern/diff/D5705442/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
Looks good, please wait for tests to pass.

xush6528 committed:
Update on "Change type of timeoutFutures_ key to time_point instead of duration"


Make `ProcessGroupAgent::pollTimedOutRPCs` code more conventional.

- Use `std::chrono::time_point` to represent `endTime` instead of `std::chrono::duration`.
- Replace `std::condition_variable::wait_for(lock, endTime)` with `std::condition_variable::wait_until(lock, endTime)`.
- Remove the unnecessary `::getRPCRemainingTime()`.

Differential Revision: [D5705442](https://our.internmc.facebook.com/intern/diff/D5705442/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -13,6 +13,12 @@ namespace torch {
+using steady_clock_time_point =
+    std::chrono::time_point<std::chrono::steady_clock>;
+
+constexpr steady_clock_time_point kInfiniteTimeoutTimePoint =
```
why do you need this in the .h?

xush6528 replied:
@jjlilley Wanted to put it in the class declaration.

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -13,6 +13,12 @@ namespace torch {
 namespace distributed {
 namespace rpc {
 
+using steady_clock_time_point =
```
can this be a class-scope using, rather than exposed throughout all distributed::rpc scope?

xush6528 replied:
@jjlilley Let me put int under ProcessGroupAgent class declaration.

xush6528 committed:
Update on "Change type of timeoutFutures_ key to time_point instead of duration"


Make `ProcessGroupAgent::pollTimedOutRPCs` code more conventional.

- Use `std::chrono::time_point` to represent `endTime` instead of `std::chrono::duration`.
- Replace `std::condition_variable::wait_for(lock, endTime)` with `std::condition_variable::wait_until(lock, endTime)`.
- Remove the unnecessary `::getRPCRemainingTime()`.

Differential Revision: [D5705442](https://our.internmc.facebook.com/intern/diff/D5705442/)

[ghstack-poisoned]

jjlilley reviewed (approved):
LGTM, pending circleci, thanks!

xush6528 committed:
Update on "Change type of timeoutFutures_ key to time_point instead of duration"


Make `ProcessGroupAgent::pollTimedOutRPCs` code more conventional.

- Use `std::chrono::time_point` to represent `endTime` instead of `std::chrono::duration`.
- Replace `std::condition_variable::wait_for(lock, endTime)` with `std::condition_variable::wait_until(lock, endTime)`.
- Remove the unnecessary `::getRPCRemainingTime()`.

Differential Revision: [D5705442](https://our.internmc.facebook.com/intern/diff/D5705442/)

[ghstack-poisoned]

xush6528 committed:
Update on "Change type of timeoutFutures_ key to time_point instead of duration"


Make `ProcessGroupAgent::pollTimedOutRPCs` code more conventional.

- Use `std::chrono::time_point` to represent `endTime` instead of `std::chrono::duration`.
- Replace `std::condition_variable::wait_for(lock, endTime)` with `std::condition_variable::wait_until(lock, endTime)`.
- Remove the unnecessary `::getRPCRemainingTime()`.

Differential Revision: [D5705442](https://our.internmc.facebook.com/intern/diff/D5705442/)

[ghstack-poisoned]

xush6528 committed:
Update on "Change type of timeoutFutures_ key to time_point instead of duration"


Make `ProcessGroupAgent::pollTimedOutRPCs` code more conventional.

- Use `std::chrono::time_point` to represent `endTime` instead of `std::chrono::duration`.
- Replace `std::condition_variable::wait_for(lock, endTime)` with `std::condition_variable::wait_until(lock, endTime)`.
- Remove the unnecessary `::getRPCRemainingTime()`.

Differential Revision: [D5705442](https://our.internmc.facebook.com/intern/diff/D5705442/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 20c3823:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=20c3823db973579c386ceb6bcab53143cf9ccc00).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3913570) pytorch-linux-xenial-py3-clang5-android-ndk-r19c-gradle-build (1/1)
**Step:** "pytorch android gradle build" ([details](https://dr.pytorch.org/build-details.html?build_id=72715497))
```
sys	0m0.159s
++ docker run --cap-add=SYS_PTRACE --security-opt seccomp=unconfined -t -d -w /var/lib/jenkins 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:405-20c3823db973579c386ceb6bcab53143cf9ccc00-android-x86_32
+ export id_x86_32=40c9aeb6161cf211c89cfc8d390a3bd6c7f3b436127ab1308d9bf866990f493a
+ id_x86_32=40c9aeb6161cf211c89cfc8d390a3bd6c7f3b436127ab1308d9bf866990f493a
+ export 'COMMAND=((echo "source ./workspace/env" && echo "sudo chown -R jenkins workspace") | docker exec -u jenkins -i "$id_x86_32" bash) 2>&1'
+ COMMAND='((echo "source ./workspace/env" && echo "sudo chown -R jenkins workspace") | docker exec -u jenkins -i "$id_x86_32" bash) 2>&1'
+ echo '((echo' '"source' './workspace/env"' '&&' echo '"sudo' chown -R jenkins 'workspace")' '|' docker exec -u jenkins -i '"$id_x86_32"' 'bash)' '2>&1'
+ unbuffer bash ./command.sh
+ ts
+ docker pull 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:405-20c3823db973579c386ceb6bcab53143cf9ccc00-android-arm-v7a
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:405-20c3823db973579c386ceb6bcab53143cf9ccc00-android-arm-v7a not found
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).
